enhancements/console/pinned-resources.md

```diff
@@ -0,0 +1,190 @@
+
+## Motivation
+
+Some customers want to provide default pre-pinned resources for all users.
+
+### User Stories
+
+#### Story 1
+
+Customer wants to customize the developer perspective navigation by providing pre-pinned resources to users who do not have their navigation items customized.
+
+As a cluster admin, I want to provide default pre-pinned resources for all users so that the developers who donot have their navigation items customized will see these default pinned resources.
+
+### Goals
+
+- Allow cluster admins to define default pre-pinned resources on the dev perspective navigation
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+We can extend the existing `operator.openshift.io/v1` / `Console CRD`.
+We can add `spec.customization.perspectives.pinnedResources` which lists default pre-pinned resources
+
+### Workflow Description
+
+To add default pre-pinned resources:
+
+```text
+├── spec
+│   ├── customization
+|       ├── perspectives
+|           ├── id
+|           ├── ...
+|           ├── pinnedResources
+|               └── group
+                └── version
+                └── resource
+└── ...
+```
+
+Each perspective conforms to the following schema :
+
+- `pinnedResources: operatorv1.PinnedResourceReference[]` , defines a list of pinned resources. The default pinned resources are contributed by the console UI extension plugin.  
+- `group: string` , defines the group of the resource.
+- `version: string` , defines the version of the resource.
+- `resource: string` , defines the type of the resource. It is the plural form of the resource kind in lowercase.
+
+If the `pinnedResources` list is empty, the default pinned resources contributed by the console UI extension plugin will be shown.
+If the `pinnedResources` list is non-empty, only the listed items will be shown if the user hasn't already customized the navigation items.
+If the `group`, `version` or `resource` is incorrect, the resource to be pinned will be ignored.
+
+Note: The Administrator perspective (id="admin") doesn't support pinned resources at the moment and can be excluded via CEL validation.
+
+
+### API Extensions
+
+Extend the existing `operator.openshift.io/v1` / `Console CRD` by adding `spec.customization.perspectives.pinnedResources`
+
+So the example YAML should look like this:
+
+To set default pre-pinned resources
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata:
+  name: cluster
+  ...
+spec:
+  customization:
+    perspectives:
+    - id: 'dev'
+      ...
+      pinnedResources:
+      - group: apps
+        version: v1
+        resource: deployments
+      - group: ""
+        version: v1
+        resource: configmaps
+    - ...
+  ...
+```
+
+### Risks and Mitigations
+
+**Risk 1:** A cluster admin configures an invalid combination of API/Group/Version
+Mitigation: Invalid combination of group, version and resource will not be shown to users. The cluster admin interface can show them with a warning.
```

Comment:
Where is the warning displayed?

I'd probably only show the warning to admins who have permission to fix the problem.

Reply:
The warning is shown as an Alert on submission of the yaml

Reply:
we have added a regex to validate if the values entered for resource, version, and group are correct

Reply:
We can show the warning just in the "console cluster configuration" UI, where the admin can configure it.

We now the available CRDs there and can show configured entries that will not shown to the user because there is no CRD/resource available.

Like this:

```yaml
perspectives:
  id: dev
  pinnedResources:
  - resource: Secret
  - resource: Pipelinee # wrong with two e
```

The UI could show:

- (S) Secret
- (P) Pipelinee :warning: with a tooltip "Missing CRD, this pinned resource might not be shown."

Reply:
@lokanandaprabhu please take care of invalid configurations in the UI and show them anyway with such a warning.

Reply:
This gets a bit tricky since we do API discovery in the background, but I wouldn't show a permanent warning to users who don't have a way to fix the problem. Normal users won't be able to edit this config. It would be better to hide the nav item IMO.

I guess there is the case where the user pins a custom resource themselves and then the CRD is removed, but we don't have a way to tell, correct?

Reply:
when the user pins a custom resource themselves and the CRD is removed, the pinned resource is no longer shown on the navigation.

After discussing this with @jerolimov, we would show a warning alert icon on the customization form for the unrecognized resource, on the navigation the user would only see the recognized ones.

Reply:
> After discussing this with @jerolimov, we would show a warning alert icon on the customization form for the unrecognized resource, on the navigation the user would only see the recognized ones.

Sounds good to me.

We should clarify again what we mean by "incorrect" here. "Missing" might be a better term.